### PR TITLE
[WIP] Handle sanitation and gfx-hal update

### LINKED ISSUE #103: Creating zeroed backend types might crash
In some places (eg. [`gfxBindBufferMemory`](https://github.com/gfx-rs/portability/blob/master/libportability-gfx/src/impls.rs#L1147)) we create a zeroed type of the backend type and `mem::replace` it. If the backend type has a `Drop` impl rust will try to drop this zeroed type which might lead to bad things.

I'm not really sure what the intent of the `mem::replace` call is, does anyone know?

### PR DESCRIPTION
**Pull request by kvark**
Depends on https://github.com/gfx-rs/gfx/pull/2195
Depends on https://github.com/gfx-rs/metal-rs/pull/59
Fixes #103

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -4,3 +4,7 @@ members = [
     "libportability-gfx",
     "libportability-icd",
 ]
+
+[patch.crates-io]
+objc = { path = "../rust-objc" }
+metal-rs = { path = "../metal-rs" }
```
Comment (msiglreith) [inline Cargo.toml]: nit: newline

File: libportability-gfx/Cargo.toml [hunk 2]
```diff
@@ -28,25 +29,25 @@ version = "0.3"
 optional = true
 
 [dependencies.gfx-hal]
-git = "https://github.com/gfx-rs/gfx"
-#path = "../../gfx/src/hal"
+#git = "https://github.com/gfx-rs/gfx"
+path = "../../gfx/src/hal"
```
Comment (msiglreith) [inline libportability-gfx/Cargo.toml]: should be changed back at the end

File: libportability-gfx/src/impls.rs [hunk 3]
```diff
@@ -1105,9 +1107,7 @@ pub extern "C" fn gfxBindBufferMemory(
     memory: VkDeviceMemory,
     memoryOffset: VkDeviceSize,
 ) -> VkResult {
-    let temp = unsafe { mem::zeroed() };
-
-    *buffer = match mem::replace(&mut *buffer, temp) {
+    let new = match mem::replace(&mut *buffer, unsafe { mem::zeroed() }) {
```
Comment (msiglreith) [inline libportability-gfx/src/impls.rs]: is the `mem::replace` still needed?

File: libportability-gfx/src/handle.rs [hunk 4]
```diff
@@ -1,13 +1,14 @@
...
-#[cfg(feature = "nightly")]
 use std::sync::{Arc, Mutex};
 
+#[cfg(feature = "nightly")]
+use hal::backend::FastHashMap;
+
 #[cfg(feature = "nightly")]
 lazy_static! {
-    static ref REGISTRY: Arc<Mutex<HashMap<usize, &'static str>>> = Arc::new(Mutex::new(HashMap::new()));
+    static ref REGISTRY: Arc<Mutex<FastHashMap<usize, &'static str, >>> = Arc::new(Mutex::new(FastHashMap::default()));
```
Comment (grovesNL) [inline libportability-gfx/src/handle.rs]: nit: comma

File: libportability-gfx/src/impls.rs [hunk 5]
```diff
@@ -2450,7 +2466,8 @@ pub extern "C" fn gfxUpdateDescriptorSets(
     let write_infos = unsafe {
         slice::from_raw_parts(pDescriptorWrites, descriptorWriteCount as _)
     };
-    let mut writes = Vec::new(); //TODO: avoid allocation here and below
+    //Warning: trying to do everything in an iterator causes weird crashes...
```
Comment (grovesNL) [inline libportability-gfx/src/impls.rs]: nit: maybe leave as a TODO then, or we could open an issue

**Comment by kvark** [comment]
Nits are addressed now, thanks for the reviews!
@msiglreith the replace is still needed there, unless you can find a better way.

Bors r=msiglreith,grovesNL

**Comment by kvark** [comment]
Bors r=msiglreith,grovesNL

**Pull request merged**